src/tree/binary/merkle_tree.cairo

```diff
@@ -0,0 +1,158 @@
+    //}
+    match error {
+        ErrorCodes::NoError => (),
+        ErrorCodes::InvalidNumberOfSideNodes => { return (false, error); },
+        ErrorCodes::KeyNotInTree => { return (false, error); },
+        ErrorCodes::InvalidNumberOfLeavesInProof => { return (false, error); },
+        ErrorCodes::UnexpectedInnerHashes => { return (false, error); },
+        ErrorCodes::ExpectedAtLeastOneInnerHash => { return (false, error); },
+    }
+
+    return (computed_hash == root, ErrorCodes::NoError);
+}
+
+fn compute_root_hash(
+    key: u256, num_leaves: u256, leaf_hash: u256, side_nodes: Span<u256>
+) -> (u256, ErrorCodes) {
+    if (num_leaves == 0) {
+        return (leaf_hash, ErrorCodes::InvalidNumberOfLeavesInProof);
+    }
+    if (num_leaves == 1) {
+        if (side_nodes.len() != 0) {
+            return (leaf_hash, ErrorCodes::UnexpectedInnerHashes);
+        }
+        return (leaf_hash, ErrorCodes::NoError);
+    }
+    if (side_nodes.len() == 0) {
+        return (leaf_hash, ErrorCodes::ExpectedAtLeastOneInnerHash);
+    }
+
+    let num_left: u256 = get_split_point(num_leaves);
+    let side_nodes_left: Array<u256> = slice(side_nodes, 0, side_nodes.len().into() - 1);
+    if (key < num_left) {
+        let (left_hash, error) = compute_root_hash(
+            key, num_left, leaf_hash, side_nodes_left.span()
+        );
+        //match error {
+        //    ErrorCodes::NoError => (),
+        //    _ => { return (leaf_hash, error); }
+        //}
+        match error {
+            ErrorCodes::NoError => (),
+            ErrorCodes::InvalidNumberOfSideNodes => { return (leaf_hash, error); },
+            ErrorCodes::KeyNotInTree => { return (leaf_hash, error); },
+            ErrorCodes::InvalidNumberOfLeavesInProof => { return (leaf_hash, error); },
+            ErrorCodes::UnexpectedInnerHashes => { return (leaf_hash, error); },
+            ErrorCodes::ExpectedAtLeastOneInnerHash => { return (leaf_hash, error); },
+        }
+        return (node_digest(left_hash, *side_nodes.at(side_nodes.len() - 1)), ErrorCodes::NoError);
+    }
+
+    let (right_hash, error) = compute_root_hash(
+        key - num_left, num_leaves - num_left, leaf_hash, side_nodes_left.span()
+    );
+    //match error {
+    //    ErrorCodes::NoError => (),
+    //    _ => { return (leaf_hash, error); }
+    //}
+    match error {
+        ErrorCodes::NoError => (),
+        ErrorCodes::InvalidNumberOfSideNodes => { return (leaf_hash, error); },
+        ErrorCodes::KeyNotInTree => { return (leaf_hash, error); },
+        ErrorCodes::InvalidNumberOfLeavesInProof => { return (leaf_hash, error); },
+        ErrorCodes::UnexpectedInnerHashes => { return (leaf_hash, error); },
+        ErrorCodes::ExpectedAtLeastOneInnerHash => { return (leaf_hash, error); },
+    }
+    return (node_digest(*side_nodes.at(side_nodes.len() - 1), right_hash), ErrorCodes::NoError);
+}
+
+fn slice(mut data: Span<u256>, begin: u256, end: u256) -> Array<u256> {
```

Comment:
Yes for Array<u128>, https://github.com/keep-starknet-strange/alexandria/blob/39d20fb71ea4f64b9fe4d749eee475eb0b77a8c5/src/bytes/src/utils.cairo#L144

Reply:
I noticed that, too, but I ended up only using Spans because it is cheaper gas-wise for what I was doing. Also, it has a slice op built in.

There are a couple of issues with the slice impl in `Alexandria` as well. I'm going to submit an issue / PR to fix them.